prog/collide.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2021 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+// Contains prog transformations that intend to trigger more races.
+
+package prog
+
+import "math/rand"
+
+// TODO: add tests once the contents is more or less finalized.
+
+// Ensures that, if a detached call produces a resource, then
+// it is [distanced] from a call consuming the resource at least
+// by one non-detached call.
+// This does not give 100% guarantee that the detached call finishes
+// by that time, but hopefully this is enough for most cases.
+func AssignRandomDetached(origProg *Prog, rand *rand.Rand) *Prog {
+	prog := origProg.Clone()
```

Comment:
I am thinking if this function can be made simpler by scanning calls backwards.
Then we can avoid detaching syscall if the outputs are used by the next syscall and remove the undo mechanism. And also we could stop scanning once we detach 6 syscalls.